### ISSUE TITLE
No clear example in doc for using MaskablePPO in custom env

### ISSUE THREAD

**Issue by vahidqo** [Stable-Baselines-Team/stable-baselines3-contrib#72]
Hi all,
I think this is a problem that there isn't any specific and clear code example for the "masking" part in MaskablePPO. So could you please provide a clear example of how to use "masking" in MaskablePPO? or at least provide a very simple example of a custom environment using MaskablePPO. I can help you by providing a simple example.
Thanks

**Comment by araffin** [comment]
Hello,
have you followed the two examples in https://sb3-contrib.readthedocs.io/en/master/modules/ppo_mask.html#example ?
what have you tried so far that didn't work?

**Comment by araffin** [comment]
source code of the demo env can be found here : https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/75b2de139927da26d5871aef9fd839632f73b296/sb3_contrib/common/envs/invalid_actions_env.py#L35

it's just doing what is written in the doc.

**Comment by vahidqo** [comment]
what I have learned so far is that for masking action, there must be a function that inputs the current state and returns invalid actions in that specific state but for example, there isn't any state as input. for example, in the following link, the invalid actions are fixed numbers that do not make any sense because invalid actions vary depending base on the current state.
https://sb3-contrib.readthedocs.io/en/master/modules/ppo_mask.html#example

**Comment by araffin** [comment]
if you look at the current implementation, the action mask is more general, it depends on the state of the environment (not only the state of the agent), and you have access to it because it is a class method.

**Comment by vahidqo** [comment]
that's true but there isn't any specific structure for the env.valid_action_mask(). Is it True/ False? or it's 0/1?

**Comment by araffin** [comment]
the doc says "True if the action is valid, False otherwise", and type hint says it is a numpy array, so numpy array of booleans.

**Comment by vahidqo** [comment]
Many thanks for all of your answers. They were very helpful. In the end, could you please provide a very simple custom environment using MaskablePPO?

**Comment by araffin** [comment]
I already provided it here https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/issues/72#issuecomment-1140867998 ;)

that what we use in our tests

https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/75b2de139927da26d5871aef9fd839632f73b296/sb3_contrib/common/envs/invalid_actions_env.py#L8-L36

(now I realize the first type int is wrong, we should update it to `List[bool]` instead of `List[int]` even though it should work the same thanks to safe casting)

**Comment by vahidqo** [comment]
Thanks for the replies. I did like examples but after a few times of training, the following error occurs:
`ValueError                                Traceback (most recent call last)
[<ipython-input-36-db928b74aa26>](https://localhost:8080/#) in <module>()
----> 1 model.learn(25000)

6 frames
[/usr/local/lib/python3.7/dist-packages/torch/distributions/distribution.py](https://localhost:8080/#) in __init__(self, batch_shape, event_shape, validate_args)
     54                 if not valid.all():
     55                     raise ValueError(
---> 56                         f"Expected parameter {param} "
     57                         f"({type(value).__name__} of shape {tuple(value.shape)}) "
     58                         f"of distribution {repr(self)} "

ValueError: Expected parameter probs (Tensor of shape (64, 11664)) of distribution MaskableCategorical(probs: torch.Size([64, 11664]), logits: torch.Size([64, 11664])) to satisfy the constraint Simplex(), but found invalid values:
tensor([[4.8738e-05, 7.6623e-05, 8.9728e-05,  ..., 8.4073e-05, 8.4069e-05,
         7.1501e-05],
        [4.8605e-05, 7.6439e-05, 8.9702e-05,  ..., 8.4113e-05, 8.4097e-05,
         7.1714e-05],
        [5.3171e-05, 7.8171e-05, 8.9898e-05,  ..., 8.4571e-05, 8.4597e-05,
         7.3677e-05],
        ...,
        [5.1586e-05, 7.7174e-05, 8.9807e-05,  ..., 8.4648e-05, 8.4637e-05,
         7.4600e-05],
        [4.9294e-05, 7.6707e-05, 8.9838e-05,  ..., 8.4342e-05, 8.4332e-05,
         7.2955e-05],
        [4.9261e-05, 7.6919e-05, 8.9834e-05,  ..., 8.4136e-05, 8.4123e-05,
         7.1582e-05]], grad_fn=<SoftmaxBackward0>)`

**Comment by araffin** [comment]
Closing as initial question/request was answered.

> Thanks for the replies. I did like examples but after a few times of training, the following error occurs:

this goes into the technical support category, which we don't do (see instructions in our issues/in the SB3 readme).
However, if you find a bug, please an issue with a **minimal** code example to reproduce it.

**Comment by koliber31** [comment]
> Thanks for the replies. I did like examples but after a few times of training, the following error occurs: `
> 
> > ValueError                                Traceback (most recent call last)
> >  in ()
> > ----> 1 model.learn(25000)
> > 6 frames
> > [/usr/local/lib/python3.7/dist-packages/torch/distributions/distribution.py](https://localhost:8080/#) in **init**(self, batch_shape, event_shape, validate_args)
> > 54                 if not valid.all():
> > 55                     raise ValueError(
> > ---> 56                         f"Expected parameter {param} "
> > 57                         f"({type(value).**name**} of shape {tuple(value.shape)}) "
> > 58                         f"of distribution {repr(self)} "
> > ValueError: Expected parameter probs (Tensor of shape (64, 11664)) of distribution MaskableCategorical(probs: torch.Size([64, 11664]), logits: torch.Size([64, 11664])) to satisfy the constraint Simplex(), but found invalid values:
> > tensor([[4.8738e-05, 7.6623e-05, 8.9728e-05,  ..., 8.4073e-05, 8.4069e-05,
> > 7.1501e-05],
> > [4.8605e-05, 7.6439e-05, 8.9702e-05,  ..., 8.4113e-05, 8.4097e-05,
> > 7.1714e-05],
> > [5.3171e-05, 7.8171e-05, 8.9898e-05,  ..., 8.4571e-05, 8.4597e-05,
> > 7.3677e-05],
> > ...,
> > [5.1586e-05, 7.7174e-05, 8.9807e-05,  ..., 8.4648e-05, 8.4637e-05,
> > 7.4600e-05],
> > [4.9294e-05, 7.6707e-05, 8.9838e-05,  ..., 8.4342e-05, 8.4332e-05,
> > 7.2955e-05],
> > [4.9261e-05, 7.6919e-05, 8.9834e-05,  ..., 8.4136e-05, 8.4123e-05,
> > 7.1582e-05]], grad_fn=)
> 
> `

Did you menage to solve this error. I'm facing the same problem.
